src/masternodeman.cpp

```diff
@@ -844,7 +848,26 @@ void CMasternodeMan::UncacheBlockHash(const CBlockIndex* pindex)
 
 uint256 CMasternodeMan::GetHashAtHeight(int nHeight) const
 {
-    return cvLastBlockHashes.Get(nHeight);
+    // return zero if outside bounds
+    if (nHeight < 0) {
+        LogPrint(BCLog::MASTERNODE, "%s: Negative height. Returning 0\n",  __func__);
+        return UINT256_ZERO;
+    }
+    int nCurrentHeight = GetBestHeight();
+    if (nHeight > nCurrentHeight) {
+        LogPrint(BCLog::MASTERNODE, "%s: height %d over current height %d. Returning 0\n",
+                __func__, nHeight, nCurrentHeight);
+        return UINT256_ZERO;
+    }
+
+    if (nHeight > nCurrentHeight - (int) CACHED_BLOCK_HASHES) {
+        // Use cached hash
+        return cvLastBlockHashes.Get(nHeight);
+    } else {
+        // Use chainActive
+        LOCK(cs_main);
+        return chainActive[nHeight]->GetBlockHash();
```

Comment:
Note for the future: this `cs_main` could be a potential deadlock in `GetNextMasternodeInQueueForPayment`, where we are locking the internal `CMasternodeMan::cs` first.

Reply:
Yeah. It should be addressed in the future (even before deterministic masternodes, `GetNextMasternodeInQueueForPayment` can be refactored,  rearranging its locks properly and removing the redundancies). 
But, at least with this patch, it's only on blocks deeper than 200.
Currently on master we lock it for **every** block (even twice inside `CalculateScore`).